tests/std/tests/P2443R1_views_chunk_by/test.cpp

```diff
@@ -231,6 +231,24 @@ constexpr void instantiation_test() {
 #endif // TEST_EVERYTHING
 }
 
+struct Bool {
+    Bool() {}
+    Bool(const Bool&) = delete;
+    Bool& operator!() {
+        return *this;
+    }
+    operator bool() {
+        return true;
+    }
+};

```

Comment:
This type does not model `boolean-testable`, so it cannot be used as the result of binary predicate.

[`[concept.booleantestable]/7`](http://eel.is/c++draft/concept.booleantestable) says:
> Let `e` be an expression such that `decltype((e))` is `T`. `T` models *`boolean-testable`* only if `bool(e) == !bool(!e)`.

[https://godbolt.org/z/oGeefGW6G](https://godbolt.org/z/oGeefGW6G)

Reply:
I hope I can get away with it because the function is not executed.

If conformance to the semantic requirements is desired, I can make `operator!` return `std::cout`. 😈

Reply:
The name `Bool` looks terrible to me. Should it be `TrueLike`?

Reply:
> The name `Bool` looks terrible to me. Should it be `TrueLike`?

Agreed that `Bool` is not a good name, but I think `TrueLike` would be misleading. The test only needs a *`boolean-testable`* type whose `operator!` returns a reference to a non-copy-constructible type. It doesn't matter whether its `operator bool` returns `true` or not.

Reply:
Is `NonCopyableBool?` better?

Reply:
I want to use `PinnedSomething` for short. I think the following types have desired properties and model _`boolean-testable`_, but maybe too wordy.

```C++
class PinnedBool;

class PinnedFalse {
private:
    PinnedFalse() = default;
    PinnedFalse(const PinnedFalse&) = delete;
    PinnedFalse& operator=(const PinnedFalse&) = delete;

    friend class PinnedBool;

public:
    operator bool() {
        return false;
    }

    PinnedBool& operator!();
};

class PinnedBool : public PinnedFalse {
public:
    PinnedBool() = default;
    PinnedBool(const PinnedBool&) = delete;
    PinnedBool& operator=(const PinnedBool&) = delete;

    operator bool() {
        return true;
    }

    PinnedFalse& operator!() {
        return static_cast<PinnedFalse&>(*this);
    }
};

inline PinnedBool& PinnedFalse::operator!() {
    return static_cast<PinnedBool&>(*this);
}
```

Reply:
`ImmoBool`! (As a rule, ignore naming suggestions from me.)

Reply:
I believe I've addressed the concerns in this discussion by moving `Bool` into the function that uses it and adding a comment that it should not be used in runtime code. Feel free to reopen discussion if you don't agree.